### PR TITLE
Cropping mechanism

### PR DESCRIPTION
**Pull request by GGreenix**
Added cropping mechanism, the client side was done by someone else and I just continued its work and finished the core mechanism, the dynamic cropping is working as well and can be transmitted from a NT table via setDynamicCropping method of a PhotonCamera object.

Some stuff still needs to be polished but overall I want to understand if there are certain stuff that needs to be changed before I continue.

thanks in advance looking forward to see it in production :D

### CODE REVIEW

File: photon-client/package.json [hunk 1]
```diff
@@ -1,50 +0,0 @@
-{
-  "name": "photonclient",
```
Comment (mcm001) [inline photon-client/package.json:2]: Why was all of photon client deleted?

**Comment by mcm001** [comment]
Supersedes https://github.com/PhotonVision/photonvision/pull/1446

File: photon-client/package.json (hunk 1 above)
Comment (GGreenix) [reply photon-client/package.json:2]: I had to branch out from the new master so I copied the code from the other guys branch and the one I worked on some to the new one, its overall the same just with the cropping tab

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/AprilTagDetectionPipe.java [hunk 2]
```diff
@@ -59,7 +59,6 @@ protected List<AprilTagDetection> process(CVMat in) {
     public void setParams(AprilTagDetectionPipeParams newParams) {
         if (this.params == null || !this.params.equals(newParams)) {
             m_detector.setConfig(newParams.detectorParams);
-            m_detector.setQuadThresholdParameters(newParams.quadParams);
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/AprilTagDetectionPipe.java:62]: sus

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/AprilTagDetectionPipeParams.java [hunk 3]
```diff
@@ -23,15 +23,10 @@
 public class AprilTagDetectionPipeParams {
     public final AprilTagFamily family;
     public final AprilTagDetector.Config detectorParams;
-    public final AprilTagDetector.QuadThresholdParameters quadParams;
 
-    public AprilTagDetectionPipeParams(
-            AprilTagFamily tagFamily,
-            AprilTagDetector.Config config,
-            AprilTagDetector.QuadThresholdParameters quadParams) {
+    public AprilTagDetectionPipeParams(AprilTagFamily tagFamily, AprilTagDetector.Config config) {
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/AprilTagDetectionPipeParams.java:27]: revert these changes?

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/DrawCalibrationPipe.java [hunk 4]
```diff
@@ -58,11 +56,6 @@ protected Void process(Pair<Mat, List<TrackedTarget>> in) {
         int i = 0;
         for (var target : in.getRight()) {
             for (var c : target.getTargetCorners()) {
-                if (c.x < 0 || c.y < 0) {
-                    // Skip if the corner is less than zero
-                    continue;
-                }
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/DrawCalibrationPipe.java:64]: Why this diff?

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/DrawCalibrationPipe.java [hunk 5]
```diff
@@ -84,11 +77,9 @@ protected Void process(Pair<Mat, List<TrackedTarget>> in) {
 
     public static class DrawCalibrationPipeParams {
         private final FrameDivisor divisor;
-        public boolean drawAllSnapshots;
 
-        public DrawCalibrationPipeParams(FrameDivisor divisor, boolean drawSnapshots) {
+        public DrawCalibrationPipeParams(FrameDivisor divisor) {
             this.divisor = divisor;
-            this.drawAllSnapshots = drawSnapshots;
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/DrawCalibrationPipe.java:91]: accidental changes here? restore?

File: photon-core/src/main/java/org/photonvision/vision/pipeline/AdvancedPipelineSettings.java [hunk 6]
```diff
@@ -90,6 +92,60 @@ public AdvancedPipelineSettings() {
...
+    public int static_height = Integer.MAX_VALUE;
+
+    private final NTDataChangeListener dynamicCropListener;
+    private IntegerArraySubscriber dynamicCropRequest;
+    private final IntegerArrayPublisher ledModeState;
+
+    public int dynamic_x = 0;
+    public int dynamic_y = 0;
+    public int dynamic_width = Integer.MAX_VALUE;
+    public int dynamic_height = Integer.MAX_VALUE;
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipeline/AdvancedPipelineSettings.java:108]: Does this make sense to serialize?

More in general I'm not sure how I feel about dynamic crop via NT -- it's pretty perpendicular to all our other stuff to date.

File: photon-core/src/main/java/org/photonvision/vision/pipeline/AprilTagPipeline.java [hunk 7]
```diff
@@ -69,6 +75,11 @@ public AprilTagPipeline(AprilTagPipelineSettings settings) {
 
     @Override
     protected void setPipeParamsImpl() {
+        Rect staticCrop = settings.getStaticCrop();
+
+        staticCropPipe.setParams(staticCrop);
+        staticCropPipe.setDynamicRect(settings.getDynamicCrop());
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipeline/AprilTagPipeline.java:81]: >static crop
>setdynamic

hmmm

File: photon-core/src/main/java/org/photonvision/vision/pipeline/AprilTagPipeline.java [hunk 8]
```diff
@@ -88,20 +99,7 @@ protected void setPipeParamsImpl() {
...
-        quadParams.minClusterPixels = 5;
-        // these are the same as the values in WPILib 2025
-        // setting them here to prevent upstream changes from changing behavior of the detector
-        quadParams.maxNumMaxima = 10;
-        quadParams.criticalAngle = 45 * Math.PI / 180.0;
-        quadParams.maxLineFitMSE = 10.0f;
-        quadParams.minWhiteBlackDiff = 5;
-        quadParams.deglitch = false;
-
-        aprilTagDetectionPipe.setParams(
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipeline/AprilTagPipeline.java:103]: accidental changes?

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/CropPipe.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        //     return in;
+        // }
+        Rect proccesor = false ? this.params : dynamiRect;
+
+        int x = MathUtil.clamp(proccesor.x, 0, mat.width());
+        int y = MathUtil.clamp(proccesor.y, 0, mat.height());
+        int width = MathUtil.clamp(proccesor.width, 0, mat.width() - x);
+        int height = MathUtil.clamp(proccesor.height, 0, mat.height() - y);
+
+        return new CVMat(mat.submat(y, y + height, x, x + width));
```
Comment (mcm001) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/CropPipe.java:49]: The submat here is just a submat of the overall data array, and does not actually own the native memory. but it still needs to be released. So we need to make sure that we `release` both after we are done with them. I'd argue we should release the cropped Mat immediately after Apriltag detection, since we don't use it anywhere else. See this example:


```java
   @Test
    public void testRelease() {
        var pipe = new CropPipe();
        pipe.setParams(new Rect(10, 10, 50, 50));

        var frameProvider =
                new FileFrameProvider(
                        TestUtils.getApriltagImagePath(TestUtils.ApriltagTestImages.kTag1_640_480, false),
                        TestUtils.WPI2020Image.FOV,
                        TestUtils.get2020LifeCamCoeffs(false));
        frameProvider.requestFrameThresholdType(FrameThresholdType.GREYSCALE);

        var frame = frameProvider.get();
        var out = pipe.run(frame.processedImage);

        System.out.println(frame.processedImage.getMat());
        System.out.println(out.output.getMat());

        Imgcodecs.imwrite("pre_release_full.png", frame.processedImage.getMat());
        Imgcodecs.imwrite("pre_release_crop.png", out.output.getMat());
        
        frame.release();

        System.out.println(frame.processedImage.getMat());
        System.out.println(out.output.getMat());

        Imgcodecs.imwrite("pose_releaseframe_crop.png", out.output.getMat());

        out.output.getMat().release();

        System.out.println(frame.processedImage.getMat());
        System.out.println(out.output.getMat());
    }
```

Output is:
```
// Starting state
    Mat [ 480*640*CV_8UC1, isCont=true, isSubmat=false, nativeObj=0x7ff994f80a30, dataAddr=0x7ff994fb26c0 ]
    Mat [ 50*50*CV_8UC1, isCont=false, isSubmat=true, nativeObj=0x7ff994f87910, dataAddr=0x7ff994fb3fca ]
// After releasing the full image Mat
    Mat [ 0*0*CV_8UC1, isCont=true, isSubmat=false, nativeObj=0x7ff994f80a30, dataAddr=0x0 ]
    Mat [ 50*50*CV_8UC1, isCont=false, isSubmat=true, nativeObj=0x7ff994f87910, dataAddr=0x7ff994fb3fca ]
// after releasing both
    Mat [ 0*0*CV_8UC1, isCont=true, isSubmat=false, nativeObj=0x7ff994f80a30, dataAddr=0x0 ]
    Mat [ 0*0*CV_8UC1, isCont=false, isSubmat=true, nativeObj=0x7ff994f87910, dataAddr=0x0 ]
```

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/AprilTagDetectionPipeParams.java
Comment (Juniormunk) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/AprilTagDetectionPipeParams.java:1]: This file should not have changed.

File: photon-core/src/main/java/org/photonvision/vision/pipe/impl/DrawCalibrationPipe.java
Comment (Juniormunk) [inline photon-core/src/main/java/org/photonvision/vision/pipe/impl/DrawCalibrationPipe.java:1]: This file should not have changed.

**Review by Juniormunk — COMMENTED** [review]
Looks like the robot examples contain changes as that shouldn't happen either.

File: photon-server/src/main/resources/web/index.html
Comment (crschardt) [inline photon-server/src/main/resources/web/index.html:1]: This file is autogenerated and is not supposed to be committed.

Comment (crschardt) [reply photon-server/src/main/resources/web/index.html:1]: After resetting it back to the version in master, you can run `git update-index --skip-worktree photon-server/src/main/resources/web/index.html` to prevent git from ever seeing it as changed and adding it to a commit.

File: photon-server/src/main/resources/web/index.html [hunk 10]
```diff
@@ -1 +1,14 @@
...
+    <link rel="icon" href="./favicon.ico" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Photon Client</title>
+    <script type="module" crossorigin src="./assets/index-CIrfZRaJ.js"></script>
+    <link rel="stylesheet" crossorigin href="./assets/index-Bt-HXv2B.css">
+  </head>
+  <body>
+    <div id="app"></div>
+  </body>
+</html>
```
Comment (crschardt) [inline photon-server/src/main/resources/web/index.html:14]: Accepting this suggestion will revert it to the version in master:
```suggestion
<p>UI has not been copied!</p>
```

**Comment by amquake** [comment]
Moved to #1610
